include/gul14/bit_manip.h

```diff
@@ -57,6 +57,44 @@ using BitFunctionReturnType =
         std::decay_t<T>
     >;
 
+/**
+ * An enum to determine the endianness of multi-byte scalars on the current platform.
+ *
+ * In big-endian (Motorola) order, the most significant byte is stored first, followed by
+ * the other bytes in order of decreasing significance. Little-endian platforms (e.g.
+ * Intel) store the bytes in the opposite order. There are also (historical) platforms
+ * which do not conform to either of these conventions.
+ *
+ * \code{.cpp}
+ *  if constexpr (gul14::endian::native == gul14::endian::big)
+ *      std::cout << "This is a big-endian machine!\n";
+ * \endcode
+ *
+ * This is a backport of [std::endian](https://en.cppreference.com/w/cpp/types/endian)
+ * from C++20.
+ *
+ * \see is_big_endian(), is_little_endian()
+ * \since GUL version 2.10
+ */
+enum class endian
+{
+#if defined(__BYTE_ORDER__)
+    little = __ORDER_LITTLE_ENDIAN__,
+    big    = __ORDER_BIG_ENDIAN__,
+    native = __BYTE_ORDER__
+#elif defined(_MSC_VER) && !defined(__clang__)
+    little = 0,
+    big    = 1,
+    native = little
+#else
+    #error "Don't know how to determine machine endianness on this compiler"
```

Comment:
This is bad, because that means GUL does not work / compile on platforms that do not have `__BYTE_ORDER__` or `_MSC_VER_ etc, **even if the programmer does not want to determine the endianness.**

Reply:
Well, shouldn't we simply implement a solution for that platform, then?

Reply:
I do not know. I think failing is ok, but not for all of GUL.

Self-implemented stuff looks easier than correctly done, iirc.
That's the reason these are compiler macros and not some standard library solution.

Reply:
I think failing directly when compiling GUL14 is the best option. It directly points to the place where we need to fix something. Otherwise, we end up with conditional compilation and will already have to #ifdef out our unit tests.

Reply:
I believe **if** we add a function to GUL it should work 'everywhere' not only where we want to use it. There is some higher barrier that functions in GUL need to pass that otherwise I would just wave through.

If we want to add the endian test it must be (strictly) correct; which is not easy as I already pointed out. Have a look for example at [boost's implementation](https://github.com/boostorg/predef/blob/develop/include/boost/predef/other/endian.h) which is long and splits into multiple dedicated header files.

A tongue in cheek implementation is imho not what GUL is about. Than maybe better do it in the clientlib (again), the scope there is so much smaller; it has a very limited set of platforms etc pp.

---

Whatever, I think we can rely on Meson for the problematic cases, like so:
(untested)

`src/meson.build`:

```python
if host_machine.endian() == 'big'
    add_cpp_args += [ '-DGUL_PLATFORM_BIG_ENDIAN' ]
endif
```

And here something along the lines of

```c++
#if defined(__BYTE_ORDER__)
    little = __ORDER_LITTLE_ENDIAN__,
    big    = __ORDER_BIG_ENDIAN__,
    native = __BYTE_ORDER__
#elif (defined(_MSC_VER) && !defined(__clang__)) || defined(GUL_PLATFORM_BIG_ENDIAN)
    little = 0,
    big    = 1,
    native = little
#else
    little = 0,
    big    = 1,
    native = big
#endif
```

Reply:
Well, libc++ does something very similar, so I would argue that we need not be any more generic than your average standard library:
```cpp
#  ifndef __BYTE_ORDER__
#    error                                                                                                             \
        "Your compiler doesn't seem to define __BYTE_ORDER__, which is required by libc++ to know the endianness of your target platform"
#  endif

#  if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
#    define _LIBCPP_LITTLE_ENDIAN
#  elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
#    define _LIBCPP_BIG_ENDIAN
#  endif // __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
```
See here: https://github.com/llvm/llvm-project/blob/main/libcxx/include/__config

Reply:
> Edit: Thinking about it, why not rely purely on Meson to figure it out?

Because that makes the workings of the library totally dependent on a single build system. In my book, that would be a very unfortunate choice.

Reply:
Otoh

a) All build systems have such a feature (well, except `make` maybe)
b) Why do the libraries need this information anyhow? Should they not just use `htons()`?

I think the desirability of endinesseses is sorted like this

* Avoid endianness whenever possible
* Use endian-neutral code
* Use existing conversion functions, such as htons() to convert between network byte-order and the native byte-order of the processor
* Detect endianness at run-time
* Detect endianness at configuration-time
* Detect endianness at compile-time

_[Old but still relevant source](https://sourceforge.net/p/predef/wiki/Endianness/)_

:thinking:

Reply:
I think we should tackle this if it becomes a problem. Until then keep it as simple as possible.
And if somebody tries to use an old GCC in the attempt to use C++14/17+ this will be the least of the problems.

Reply:
I could add the GUL14_BIG_ENDIAN macro as a fallback for the hypothetic non-gcc/clang/msvc compiler. If it is not set, we can default to little-endian and our tests will fail if that happens to be wrong. Is that a viable compromise?

Reply:
> b) Why do the libraries need this information anyhow? Should they not just use htons()?

That is beside the point.

Reply:
> That is beside the point.

Is it? For me at least GUL is not a container where I dump in all stuff that I want to have in some project, but selected functionality that I would tell people to use.
Contrarily I wound never ever suggest to people to use `endian`. I would give them the list above.
This feels like adding some functions to GUL that use `void*`. Just out of place.

So I want to ask the question (that you also asked in the meeting), _**is GUL the right place for this?**_
Most probably I would not let the functions in the clientlib that use `endian` pass a review also; well, I did not look at them. Then why add a function that endorses such code in a library that has the explicit target of 'clean and modern' code?